### ISSUE TITLE
Interactive code cell output effects on vertical centering/arrangement of slide content

### ISSUE THREAD

**Issue by j3mdamas** [damianavila/RISE#188]
Hi,

RISE is fantastic and allows the creation of interactive presentations, which is really great!

So, the interactive part is, for me, something really important and I am having troubles with it. This concerns the slide organization and code cell output. In a case where the output hasn't been generated, the slide is organized considering the current status, and outputs generated on-the-fly/in presentation mode may push contents outside the slide. This has to do with the vertical centering of the slide **not** being refreshed after the output is generated (I am talking about cases where the cells with output included should fit the slide). I don't know if this refresh is possible.

In alternative, if it not possible, I think that toggled output should work as a way to circumvent this. The slide mode should arrange the cells on the slide taking the content of toggled outputs into consideration. This would allow slides that are ready for the interactive generation of output, while not messing with the structure of the slide. Currently, toggled output gives the same slides as ungenerated output.

I hope my explanation is clear.

Best,
João

**Comment by fjsj** [comment]
+1

**Comment by j3mdamas** [comment]
@fjsj, is your fix working well?

**Comment by ThomasA** [comment]
I usually lay out my slides with all output evaluated to make sure it 
fits in. If I want to show my slides while generating the output on the 
fly, I then simply clear all outputs before starting the slideshow and 
then I know from when I laid them out that the output will fit.

**Comment by j3mdamas** [comment]
@ThomasA I'm pretty sure that does not work. If the output is not there, the current layout (without output considered) is vertically centered and when the output is generated, it may go out of the slide.

**Comment by damianavila** [comment]
@j3mdamas, just to be sure... you know that you can pass and `scroll: True` in the config and you get a scrollbar that lets you see the output you generated on the fly and outside the current slide view?
It is a mitigation, not a solution... but I just want to know if you are aware of that option.

**Comment by j3mdamas** [comment]
I did not know. I'm going to try it out.
But at least you understand what I mean, I am glad that I am being able to express myself :)
Do you think it's something that it's fixable? I would also like to try @fjsj fix. I am assuming it works, but I want to try it out.

**Comment by damianavila** [comment]
This is something I have to think about... and also I need to check @fjsj fix, maybe there is an answer there :wink:

**Comment by fjsj** [comment]
Hey @damianavila and @j3mdamas, my solution works on modern browsers that have [MutationObserver](https://developer.mozilla.org/en/docs/Web/API/MutationObserver). Code can probably be put in a better place and have documentation added, but the solution is this: https://github.com/vintasoftware/RISE/commit/e7cb8da9bb8eab5a34a3b00524395961c11f81a1#diff-1

For those who want to try it, obviously RISE should be installed from the dir with this code using `python setup.py install`. I've used it for presenting this talk: https://github.com/vintasoftware/pythonwat

**Comment by fjsj** [comment]
@damianavila are you open to a PR addressing this issue?

**Comment by damianavila** [comment]
Yes, of course... I am open to any PR... then we can discuss about the implementation and details, but you should not ask to put a PR :wink: Any contribution is welcomed and appreciated!

**Comment by j3mdamas** [comment]
I'll be back on this issue after holidays and when I start preparing a workshop on which I intend to use RISE a lot :wink:

**Comment by j3mdamas** [comment]
@damianavila, is this on 4.0.0b1 that is on conda?

**Comment by damianavila** [comment]
Yep, it should be there...

**Comment by j3mdamas** [comment]
Yes, it is. Sorry. It's working great for text and image output! Thanks.

Still having problems with webgl out from [nglview](https://github.com/arose/nglview)

**Comment by damianavila** [comment]
>Still having problems with webgl out from nglview

Can you open another issue for that... so we know the issue and others can think/comment about that. Thx
